components/region_cache_memory_engine/src/region_label.rs

```diff
@@ -0,0 +1,406 @@
+    pub start_at: Option<String>,
+    pub data: Vec<KeyRangeRule>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct KeyRangeRule {
+    pub start_key: String,
+    pub end_key: String,
+}
+
+// Todo: more efficient way to do this for cache use case?
+#[derive(Default)]
+pub struct RegionLabelRulesManager {
+    pub(crate) region_labels: DashMap<String, LabelRule>,
+}
+
+impl RegionLabelRulesManager {
+    pub fn add_region_label(&self, label_rule: LabelRule) {
+        let _ = self.region_labels.insert(label_rule.id.clone(), label_rule);
+    }
+
+    pub fn region_labels(&self) -> Vec<LabelRule> {
+        self.region_labels
+            .iter()
+            .map(|e| e.value().clone())
+            .collect::<Vec<_>>()
+    }
+
+    pub fn remove_region_label(&self, label_rule_id: &String) {
+        let _ = self.region_labels.remove(label_rule_id);
+    }
+
+    pub fn get_region_label(&self, label_rule_id: &str) -> Option<LabelRule> {
+        self.region_labels
+            .get(label_rule_id)
+            .map(|r| r.value().clone())
+    }
+}
+
+pub type RuleFilterFn = Arc<dyn Fn(&LabelRule) -> bool + Send + Sync>;
+
+#[derive(Clone)]
+pub struct RegionLabelService {
+    manager: Arc<RegionLabelRulesManager>,
+    pd_client: Arc<RpcClient>,
+    meta_client: Checked<Sourced<Arc<RpcClient>>>,
+    revision: i64,
+    path_suffix: Option<String>,
+    rule_filter_fn: Option<RuleFilterFn>,
+}
+
+const RETRY_INTERVAL: Duration = Duration::from_secs(1); // to consistent with pd_client
+
+pub struct RegionLabelServiceBuilder {
+    manager: Arc<RegionLabelRulesManager>,
+    pd_client: Arc<RpcClient>,
+    path_suffix: Option<String>,
+    rule_filter_fn: Option<RuleFilterFn>,
+}
+
+impl RegionLabelServiceBuilder {
+    pub fn new(
+        manager: Arc<RegionLabelRulesManager>,
+        pd_client: Arc<RpcClient>,
+    ) -> RegionLabelServiceBuilder {
+        RegionLabelServiceBuilder {
+            manager,
+            pd_client,
+            path_suffix: None,
+            rule_filter_fn: None,
+        }
+    }
+
+    pub fn path_suffix(mut self, suffix: String) -> Self {
+        self.path_suffix = Some(suffix);
+        self
+    }
+
+    pub fn rule_filter_fn(mut self, rule_filter_fn: RuleFilterFn) -> Self {
+        self.rule_filter_fn = Some(rule_filter_fn);
+        self
+    }
+
+    pub fn build(self) -> RegionLabelService {
+        RegionLabelService {
+            manager: self.manager,
+            revision: 0,
+            meta_client: Checked::new(Sourced::new(
+                Arc::clone(&self.pd_client.clone()),
+                pd_client::meta_storage::Source::RegionLabel,
+            )),
+            pd_client: self.pd_client,
+            path_suffix: self.path_suffix,
+            rule_filter_fn: self.rule_filter_fn,
+        }
+    }
+}
+
+impl RegionLabelService {
+    fn region_label_path(&self) -> String {
+        let cluster_id = self.pd_client.get_cluster_id().unwrap();
```

Comment:
the cluster_id will not be changed after cluster startup. so we can initialize it when building the service.

Reply:
done.